### ISSUE TITLE
Registering own plugins with their factories

### ISSUE THREAD

**Issue by MatthiasKuehneEllerhold** [zendframework/zend-cache#115]
How can I register my own plugins that are using a factory? Im using the zend-mvc.
Example for a view-helper:

``` php
return [
  'view_helpers' => [
    'factories' => [
      'myViewHelper' => MyViewHelperFactory::class,
    ],
  ],
];
```

If I register my plugin factory in the normal service_manager it just defaults back to the invokable factory. But my plugin has a dependency that should be injected via the constructor...
I've browsed the code but didnt find anything to make it work. Is this currently possible?

Versions in use:

```
zendframework/zend-cache                     2.7.1
zendframework/zend-config                    2.6.0
zendframework/zend-eventmanager              3.0.1
zendframework/zend-modulemanager             2.7.2              
zendframework/zend-mvc                       3.0.3              
zendframework/zend-servicemanager            3.1.1              
```

**Comment by marc-mabe** [comment]
@MatthiasKuehneEllerhold Sorry but I don't get it. Is this a question for the zend-cache component? If so please provide more information.

**Comment by MatthiasKuehneEllerhold** [comment]
Yes its for zend-cache and its integration with zend-mvc.

I want to use my own cache-plugins (e. g. a logger) that may need dependencies to work properly (e. g. a Zend/Log instance). Ideally I want to setup a factory for this plugin in order to retrieve all dependencies and set them via constructor. See the view-helper example above.

But I found no way to register own plugins that aren't invokable.

**Comment by marc-mabe** [comment]
@MatthiasKuehneEllerhold you need to add your plugin on bootstrap phase using
`Zend\Cache\StorageFactory::getPluginManager()`

Cheers

**Comment by marc-mabe** [comment]
Closing this because of no response.
In case the provided help doesn't solve your issue please feel free to re-open it again.
Thanks

**Comment by MatthiasKuehneEllerhold** [comment]
I've added this to my Module::onBootstrap() function:
```php
StorageFactory::getPluginManager()->setFactory(
    ExceptionLogger::class,
    ExceptionLoggerFactory::class
);
```
And it still uses the invokable factory for my plugin instead of the factory.

I've debugged the SM quite a bit and the plugin manager that is returned onBootstrap from `StorageFactory::getPluginManager()` is different than the one actually used later. Because the  
`StorageCacheFactory` uses the `PluginManagerLookupTrait` which sets a new plugin manager regardless if there is one already set.

After I knew that I used this and it worked.
```php
$container->get(\Zend\Cache\Storage\PluginManager::class)->setFactory(
    ExceptionLogger::class,
    ExceptionLoggerFactory::class
);
```

This is by large harder to use than every other zend-module that allows you to register factories and the like via config instead of adding them programmatically. And the singleton function `StorageFactory::getPluginManager()` is really misleading because it isn't a real singleton because the "singleton" will be overwritten regardless if one is already set.

The next big downside is that now I have to instantiate the plugin manager on every request because I *might* use the cache and its plugin. Whereas with the configuration method (from other zend modules) I'll only have to instantiate the plugin manager when I'm using the cache.
